internal/controller/api_to_config.go

```diff
@@ -255,111 +255,169 @@ func passwordForNeighbor(n v1beta1.Neighbor, passwordSecrets map[string]corev1.S
-		return frr.AllowedOut{}, err
+	prefixesForFamily := map[ipfamily.Family]sets.Set[string]{
+		ipfamily.IPv4: prefixesToAdvertiseForFamily(toAdvertise, prefixesInRouter, ipfamily.IPv4),
+		ipfamily.IPv6: prefixesToAdvertiseForFamily(toAdvertise, prefixesInRouter, ipfamily.IPv6),
 	}
-	err = setCommunitiesToAdvertisements(advsV6, communities, ipfamily.IPv6)
-	if err != nil {
-		return frr.AllowedOut{}, err
-	}
-	localPrefs, err := localPrefPrefixesToMap(toAdvertise.PrefixesWithLocalPref)
-	if err != nil {
-		return frr.AllowedOut{}, err
+
+	res := frr.AllowedOut{
+		PrefixesV4:                 make([]string, 0),
+		PrefixesV6:                 make([]string, 0),
+		LocalPrefPrefixesModifiers: make(map[string]frr.LocalPrefPrefixList),
+		CommunityPrefixesModifiers: make(map[string]frr.CommunityPrefixList),
 	}
-	err = setLocalPrefToAdvertisements(advsV4, localPrefs, ipfamily.IPv4)
-	if err != nil {
-		return frr.AllowedOut{}, err
+
+	if neighborHasIPFamily(neighbor, ipfamily.IPv4) {
+		res.PrefixesV4 = sets.List(prefixesForFamily[ipfamily.IPv4])
 	}
-	err = setLocalPrefToAdvertisements(advsV6, localPrefs, ipfamily.IPv6)
-	if err != nil {
-		return frr.AllowedOut{}, err
+	if neighborHasIPFamily(neighbor, ipfamily.IPv6) {
+		res.PrefixesV6 = sets.List(prefixesForFamily[ipfamily.IPv6])
 	}
-	res := frr.AllowedOut{
-		PrefixesV4: sortMap(advsV4),
-		PrefixesV6: sortMap(advsV6),
+
+	for _, ipFamily := range neighborIPFamilies {
+		var err error
+		res.LocalPrefPrefixesModifiers, err = prefixesWithLocalPrefToFRR(res.LocalPrefPrefixesModifiers, neighbor, toAdvertise, ipFamily, prefixesForFamily[ipFamily])
+		if err != nil {
+			return frr.AllowedOut{}, fmt.Errorf("failed to process local pref for neighbor %s, err: %w", neighbor.Name, err)
+		}
+		res.CommunityPrefixesModifiers, err = prefixesWithCommunityToFRR(res.CommunityPrefixesModifiers, neighbor, toAdvertise, ipFamily, prefixesForFamily[ipFamily])
+		if err != nil {
+			return frr.AllowedOut{}, fmt.Errorf("failed to process local pref for neighbor %s, err: %w", neighbor.Name, err)
+		}
 	}
+
 	return res, nil
 }
 
-// prefixesToMap returns two maps of prefix->OutgoingFilter (ie family, advertisement, communities), one for each family.
-// The ipv4Prefixes and ipv6Prefixes represent the "global" allowed prefixes which are the prefixes defined on the router.
-func prefixesToMap(toAdvertise v1beta1.Advertise, ipv4Prefixes, ipv6Prefixes []string) (map[string]*frr.OutgoingFilter, map[string]*frr.OutgoingFilter) {
-	resV4 := map[string]*frr.OutgoingFilter{}
-	resV6 := map[string]*frr.OutgoingFilter{}
-	if toAdvertise.Allowed.Mode == v1beta1.AllowAll {
-		for _, p := range ipv4Prefixes {
-			resV4[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: ipfamily.IPv4}
+func prefixesWithLocalPrefToFRR(toAdd map[string]frr.LocalPrefPrefixList, neighbor *frr.NeighborConfig, toAdvertise v1beta1.Advertise, ipFamily ipfamily.Family, routerPrefixes sets.Set[string]) (map[string]frr.LocalPrefPrefixList, error) {
```

Comment:
I think we can remove the toAdd here (or at least avoid re-using it inside the function)

Reply:
not sure I understand what you mean, the point here is to append items there